### PR TITLE
Added support for TaggedAddresses for services

### PR DESCRIPTION
**Pull request by evgslyusar**
Added support for TaggedAddresses for services, according to the contacts [consul API](https://www.consul.io/api-docs/catalog#servicetaggedaddresses)

### CODE REVIEW

**Review by mfkl — APPROVED** [review]
LGTM

File: Consul/Agent.cs [hunk 1]
```diff
@@ -145,6 +145,8 @@ public class AgentService
         public string[] Tags { get; set; }
         public int Port { get; set; }
         public string Address { get; set; }
+
+        public IDictionary<string, ServiceTaggedAddress> TaggedAddresses { get; set; }
```
Comment (marcin-krystianc) [inline Consul/Agent.cs:149]: I know that it is technically a dictionary, but on other hand only two keys (`lan` and `wan`) can be used. I wonder if it wouldn't be better if it is defined as a struct so there is no doubt what can and what cannot be a key?:
```
public TaggedAddresses TaggedAddresses { get; set; }
```

```
    public class TaggedAddress
    {
        public string Address { get; set; }
        public int Port { get; set; }
    }

    public class TaggedAddresses
    {
        public TaggedAddress Lan { get; set; }
        public TaggedAddress Wan { get; set; }
    }
 ```
 
 Initialisation of such struct is a little bit easier:
 ```
 var taggedAddresses = new ServiceTaggedAddresses
 {
      Lan = new ServiceTaggedAddress {Address = "127.0.0.1", Port = 80},
      Wan = new ServiceTaggedAddress {Address = "192.168.10.10", Port = 8000},
 }
 ```

File: Consul.Test/CatalogTest.cs [hunk 2]
```diff
@@ -140,6 +146,58 @@ public async Task Catalog_RegistrationDeregistration()
...
+            await _client.Catalog.Deregister(dereg);
+
+            dereg = new CatalogDeregistration
+            {
+                Datacenter = "dc1",
+                Node = "foobar",
+                Address = "192.168.10.10"
+            };
+
+            await _client.Catalog.Deregister(dereg);
```
Comment (marcin-krystianc) [inline Consul.Test/CatalogTest.cs:193]: I think that deregistration is not essential for this test. Should we remove it entirely?

Comment (mfkl) [reply Consul.Test/CatalogTest.cs:193]: Noticed that too but it's been done that way for the other tests so this PR does stick to the existing style. Fine either way

**Comment by mfkl** [comment]
Could you please update the changelog file as well?

File: Consul/Agent.cs (hunk 1 above)
Comment (evgslyusar) [reply Consul/Agent.cs:149]: Yes, WAN and LAN are specified as keys but it is not the restriction, we can use any keys. For example, we can use 'Internal', 'external' keys. Moreover, in a service registration process via a configuration file, there are no restrictions on keys.

the example below works amazing:

```
          var svcID = KVTest.GenerateTestKeyName();
          var registration = new AgentServiceRegistration
            {
                Name = svcID,
                Port = 8000,
                TaggedAddresses = new Dictionary<string, ServiceTaggedAddress>
                {
                    {"lan", new ServiceTaggedAddress {Address = "127.0.0.1", Port = 80}},
                    {"wan", new ServiceTaggedAddress {Address = "192.168.10.10", Port = 8000}},
                    {"internal", new ServiceTaggedAddress {Address = "127.0.0.2", Port = 82}},
                    {"external", new ServiceTaggedAddress {Address = "192.168.10.12", Port = 8002}}
                }
            };

            await _client.Agent.ServiceRegister(registration);
```

Comment (marcin-krystianc) [reply Consul/Agent.cs:149]: All right, so let's keep it as dictionary then 👍

File: Consul.Test/CatalogTest.cs (hunk 2 above)
Comment (marcin-krystianc) [reply Consul.Test/CatalogTest.cs:193]: I think it is done this way only in the `Catalog_RegistrationDeregistration` which makes sense. I believe it is better to remove it from this newly added tests if it is not necessary.

**Review by marcin-krystianc — COMMENTED** [review]
Looks really good, I would only remove the extra code from the `Catalog_GetTaggedAddressesService` tests.

File: Consul.Test/CatalogTest.cs (hunk 2 above)
Comment (evgslyusar) [reply Consul.Test/CatalogTest.cs:188]: done

**Review by mfkl — APPROVED** [review]

File: Consul.Test/CatalogTest.cs (hunk 2 above)
Comment (marcin-krystianc) [reply Consul.Test/CatalogTest.cs:188]: @evgslyusar 
I see that You removed extra code form `Catalog_RegistrationDeregistration` method (https://github.com/G-Research/consuldotnet/pull/111/commits/aafad6f715a6b1f10b376c1e348ac0e3292d233a). I was more talking about lines 171..188 in the `Catalog_GetTaggedAddressesService` method. Can you remvoe them as well ?

**Comment by mfkl** [comment]
Just need to remove the deregistration lines and this should be good to go.

**Comment by ruxo** [comment]
Sorry, any chance this change is going to get merged?

**Comment by marcin-krystianc** [comment]
@evgslyusar,  thank you for your contribution.

**Pull request merged**